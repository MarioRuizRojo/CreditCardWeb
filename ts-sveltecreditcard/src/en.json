{
    "addcard": "Add Credit Card",
    "loading": "Loading...",
    "placeholdercardholdername": "Card Holder Name",
    "placeholdercardsnumber": "Card's Number",
    "placeholderdate": "MM/YY",
    "placeholdercvv": "CVV",
    "add": "Add",
    "required": "Required",
    "onlyletters": "Should contain only letters",
    "should16dig": "Should contain 16 digits",
    "shoulddateform": "Should have date format",
    "should3dig": "Should contain 3 digits",
    "savechanges": "Save Changes",
    "creditcards": "Credit Cards",
    "creditcardapp": "Credit Cards App",
    "selectlang": "Select language",
    "editcard": "Edit Credit Card"
}