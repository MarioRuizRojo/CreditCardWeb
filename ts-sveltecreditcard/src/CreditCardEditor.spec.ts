import CreditCardEditor from './CreditCardEditor.svelte';
import { render, fireEvent } from '@testing-library/svelte';
import userEvent from '@testing-library/user-event';
import { Observable, Subject } from 'rxjs';

import { CreditCardC } from './CreditCardC';
import { CreditCardServiceC } from './CreditCardServiceC';
import { dataCreditCardEmpty, twoCreditCards, copyPropsTwoCreditCards, comparePropsTwoCreditCards } from './fixtures/fixtures';

//language support in tests
import { locale, _, addMessages } from 'svelte-i18n';
import en from './en.json';

/**
 * Mocked serverGetCreditCards function to mock returned values
 */
let mockGetListCards: jest.Mock<any, any>;
/**
 * Mocked setNewList function to check if its called
 */
let mockServiceNewList: jest.Mock<any, any>;
/**
 * Mocked serverAddCreditCard function to check if its called
 */
let mockServiceAddCard: jest.Mock<any, any>;
/**
 * Local observable instance creditCardToEdit$ to check
 * if onEditCard in CreditCardEditor is called when creditCardToEdit$ 
 * updates
 */
let creditCardToEdit$: Subject<CreditCardC>;

/**
 * It sets up the language translation engine before running the tests
 */
beforeAll(function (): void {
    addMessages('en', en);
    locale.set('en');
});

/**
 * Mock all service functions
 */
beforeEach(function (): void {
    creditCardToEdit$ = new Subject<CreditCardC>();
    //mocks service new card to edit subscribe function
    const mockEditSubscribe: jest.Mock<any, any> = jest.fn(
        function (): Observable<CreditCardC> {
            return creditCardToEdit$.asObservable();
        }
    );
    CreditCardServiceC.prototype.getCreditCardToEdit = mockEditSubscribe;
    //mocks service new cards list to notify to observers
    mockServiceNewList = jest.fn();
    CreditCardServiceC.prototype.setNewList = mockServiceNewList;
    //mocks service add credit card
    mockServiceAddCard = jest.fn();
    CreditCardServiceC.prototype.serverAddCreditCard = mockServiceAddCard;
    //mocks service get list cards
    mockGetListCards = jest.fn(
        function (): Promise<CreditCardC[]> {
            return new Promise((resolve, reject) => resolve(twoCreditCards));
        }
    );
    CreditCardServiceC.prototype.serverGetCreditCards = mockGetListCards;
});

/**
 * It expects default values for CreditCardEditor attributes
 */
it('check empty form', function (): void {
    const rendered = render(CreditCardEditor);//empty card form
    expect(rendered).toMatchSnapshot();
});

/**
 * It updates creditCardToEdit$ and expect the first credit card 
 * of the test data set to be rendered in the form
 */
it('check if subscribe triggers the rendering of the new card to edit', function (): void {
    const rendered = render(CreditCardEditor);//empty card form
    creditCardToEdit$.next(twoCreditCards[0]);//fires subscribe with new card to edit
    //rendered has creditCard values of creditCards[0] in its variable creditCard inside the snapshot
    expect(rendered).toMatchSnapshot();
});

/**
 * It clicks submit button and expects serverAddCreditCard to have been called
 */
it('check if submit triggers service Add Credit Card', async function (): Promise<void> {
    const { getByTestId } = render(CreditCardEditor);//empty card form
    const buttonSubmit: HTMLElement = await getByTestId("submit");
    await fireEvent.click(buttonSubmit);
    expect(mockServiceAddCard).toHaveBeenCalled();
});

/**
 * It mocks serverUpdateCreditCard
 * It updates creditCardToEdit$ observable with second credit card 
 * of the test data set
 * It expects serverUpdateCreditCard to have been called with the  
 * second credit card of the test data set when submit button is clicked
 */
it('check if submit triggers service update Credit Card', async function (): Promise<void> {
    let creditCardReceivedByService: CreditCardC = new CreditCardC(dataCreditCardEmpty);
    //mocks service update credit card
    const mockServiceUpdateCard: jest.Mock<any, any> = jest.fn(
        function (creditCard1: CreditCardC): void {
            copyPropsTwoCreditCards(creditCardReceivedByService, creditCard1);
        }
    );
    CreditCardServiceC.prototype.serverUpdateCreditCard = mockServiceUpdateCard;
    const { getByTestId } = render(CreditCardEditor);//empty card form
    creditCardToEdit$.next(twoCreditCards[1]);//fires subscribe with new card to edit
    const buttonSubmit: HTMLElement = await getByTestId("submit");
    await fireEvent.click(buttonSubmit);
    expect(mockServiceUpdateCard).toHaveBeenCalled();
    expect(comparePropsTwoCreditCards(creditCardReceivedByService, twoCreditCards[1])).toBe(true);
});
/**
 * It clicks submit and expects setNewList to have been called
 */
it('check if submit triggers service new cards list to notify to observers', async function (): Promise<void> {
    const { getByTestId } = render(CreditCardEditor);//empty card form
    const buttonSubmit: HTMLElement = await getByTestId("submit");
    await fireEvent.click(buttonSubmit);
    expect(mockServiceNewList).toHaveBeenCalled();
});

/**
 * It types "123" on cvv input, blur it and expect to render a green key icon
 */
it('check if onblur triggers validations and it pass', async function (): Promise<void> {
    const rendered = render(CreditCardEditor);//empty card form
    const inputCvv: HTMLElement = await rendered.findByTestId("cvv");
    userEvent.type(inputCvv, "123");
    await fireEvent.blur(inputCvv);
    expect(rendered).toMatchSnapshot();
});

/**
 * It types "wwww" on cvv input, blur it and expect to render a red key icon 
 * and an error message
 */
it('check if onblur triggers validations and it fails', async function () {
    const rendered = render(CreditCardEditor);//empty card form
    const inputCvv: HTMLElement = await rendered.findByTestId("cvv");
    userEvent.type(inputCvv, "wwww");
    await fireEvent.blur(inputCvv);
    expect(rendered).toMatchSnapshot();
});