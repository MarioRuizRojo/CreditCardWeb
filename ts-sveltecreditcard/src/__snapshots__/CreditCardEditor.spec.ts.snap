// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check empty form 1`] = `
Object {
  "component": CreditCardEditor {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        false,
        "addcard",
        false,
        MyValidator {
          "errorsToShow": Array [
            "",
            "",
            "",
            "",
          ],
          "passed": Array [
            false,
            false,
            false,
            false,
          ],
          "requiredError": "Required",
          "validations": Array [
            MyValidation {
              "errorMessage": "Should contain only letters",
              "ruleFormat": /\\(\\[a-zA-Z\\]\\+\\)/i,
            },
            MyValidation {
              "errorMessage": "Should contain 16 digits",
              "ruleFormat": /\\(\\[0-9\\]\\{16\\}\\)/i,
            },
            MyValidation {
              "errorMessage": "Should have date format",
              "ruleFormat": /\\^\\(0\\[1-9\\]\\|1\\[0-2\\]\\)\\\\/\\(\\[0-9\\]\\{2\\}\\)\\$/i,
            },
            MyValidation {
              "errorMessage": "Should contain 3 digits",
              "ruleFormat": /\\(\\[0-9\\]\\{3\\}\\)/i,
            },
          ],
        },
        "",
        "",
        "",
        "",
        "fas fa-circle ",
        "fas fa-credit-card ",
        "fas fa-calendar-alt ",
        "fas fa-key ",
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {},
      "skip_bound": false,
      "update": [Function],
    },
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
       
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="title mb-3 svelte-180ga2g"
          >
            Add Credit Card
             
          </h5>
           
          <form>
            <div
              class="input-group input-group-lg mb-3"
              data-testid="maindivformHolderName"
            >
              <span
                class="input-group-text"
              >
                <i
                  class="fas fa-circle "
                  data-testid="iconformHolderName"
                />
              </span>
               
              <input
                class="form-control"
                data-testid="inputformHolderName"
                name="formHolderName"
                placeholder="Card Holder Name"
                type="text"
              />
               
            </div>
             
            <div
              class="input-group input-group-lg mb-3"
              data-testid="maindivformCardNumber"
            >
              <span
                class="input-group-text"
              >
                <i
                  class="fas fa-credit-card "
                  data-testid="iconformCardNumber"
                />
              </span>
               
              <input
                class="form-control"
                data-testid="inputformCardNumber"
                name="formCardNumber"
                placeholder="Card's Number"
                type="text"
              />
               
            </div>
             
            <div
              class="row"
            >
              <div
                class="col-lg-7"
              >
                <div
                  class="input-group input-group-lg mb-3"
                  data-testid="maindivformExpirationDate"
                >
                  <span
                    class="input-group-text"
                  >
                    <i
                      class="fas fa-calendar-alt "
                      data-testid="iconformExpirationDate"
                    />
                  </span>
                   
                  <input
                    class="form-control"
                    data-testid="inputformExpirationDate"
                    name="formExpirationDate"
                    placeholder="MM/YY"
                    type="text"
                  />
                   
                </div>
              </div>
               
              <div
                class="col-lg-5"
              >
                <div
                  class="input-group input-group-lg mb-3"
                >
                  <span
                    class="input-group-text"
                  >
                    <i
                      class="fas fa-key  svelte-180ga2g"
                    />
                  </span>
                   
                  <input
                    class="form-control"
                    data-testid="cvv"
                    name="formCvv"
                    placeholder="CVV"
                    type="password"
                  />
                   
                </div>
              </div>
            </div>
             
            <div
              class="d-grid gap-2"
            >
              <button
                class="btn btn-success btn-lg"
                data-testid="submit"
                disabled=""
                type="button"
              >
                <i
                  class="fas fa-database"
                />
                 
                Add
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`check if onblur triggers validations and it fails 1`] = `
Object {
  "component": CreditCardEditor {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        false,
        "addcard",
        false,
        MyValidator {
          "errorsToShow": Array [
            "",
            "",
            "",
            "Should contain 3 digits",
          ],
          "passed": Array [
            false,
            false,
            false,
            false,
          ],
          "requiredError": "Required",
          "validations": Array [
            MyValidation {
              "errorMessage": "Should contain only letters",
              "ruleFormat": /\\(\\[a-zA-Z\\]\\+\\)/i,
            },
            MyValidation {
              "errorMessage": "Should contain 16 digits",
              "ruleFormat": /\\(\\[0-9\\]\\{16\\}\\)/i,
            },
            MyValidation {
              "errorMessage": "Should have date format",
              "ruleFormat": /\\^\\(0\\[1-9\\]\\|1\\[0-2\\]\\)\\\\/\\(\\[0-9\\]\\{2\\}\\)\\$/i,
            },
            MyValidation {
              "errorMessage": "Should contain 3 digits",
              "ruleFormat": /\\(\\[0-9\\]\\{3\\}\\)/i,
            },
          ],
        },
        "",
        "",
        "",
        "wwww",
        "fas fa-circle ",
        "fas fa-credit-card ",
        "fas fa-calendar-alt ",
        "fas fa-key text-danger ",
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {},
      "skip_bound": false,
      "update": [Function],
    },
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
       
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="title mb-3 svelte-180ga2g"
          >
            Add Credit Card
             
          </h5>
           
          <form>
            <div
              class="input-group input-group-lg mb-3"
              data-testid="maindivformHolderName"
            >
              <span
                class="input-group-text"
              >
                <i
                  class="fas fa-circle "
                  data-testid="iconformHolderName"
                />
              </span>
               
              <input
                class="form-control"
                data-testid="inputformHolderName"
                name="formHolderName"
                placeholder="Card Holder Name"
                type="text"
              />
               
            </div>
             
            <div
              class="input-group input-group-lg mb-3"
              data-testid="maindivformCardNumber"
            >
              <span
                class="input-group-text"
              >
                <i
                  class="fas fa-credit-card "
                  data-testid="iconformCardNumber"
                />
              </span>
               
              <input
                class="form-control"
                data-testid="inputformCardNumber"
                name="formCardNumber"
                placeholder="Card's Number"
                type="text"
              />
               
            </div>
             
            <div
              class="row"
            >
              <div
                class="col-lg-7"
              >
                <div
                  class="input-group input-group-lg mb-3"
                  data-testid="maindivformExpirationDate"
                >
                  <span
                    class="input-group-text"
                  >
                    <i
                      class="fas fa-calendar-alt "
                      data-testid="iconformExpirationDate"
                    />
                  </span>
                   
                  <input
                    class="form-control"
                    data-testid="inputformExpirationDate"
                    name="formExpirationDate"
                    placeholder="MM/YY"
                    type="text"
                  />
                   
                </div>
              </div>
               
              <div
                class="col-lg-5"
              >
                <div
                  class="input-group input-group-lg mb-3"
                >
                  <span
                    class="input-group-text"
                  >
                    <i
                      class="fas fa-key text-danger  svelte-180ga2g"
                    />
                  </span>
                   
                  <input
                    class="form-control"
                    data-testid="cvv"
                    name="formCvv"
                    placeholder="CVV"
                    type="password"
                  />
                   
                  <div
                    style="color: red;"
                  >
                    Should contain 3 digits
                  </div>
                </div>
              </div>
            </div>
             
            <div
              class="d-grid gap-2"
            >
              <button
                class="btn btn-success btn-lg"
                data-testid="submit"
                disabled=""
                type="button"
              >
                <i
                  class="fas fa-database"
                />
                 
                Add
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`check if onblur triggers validations and it pass 1`] = `
Object {
  "component": CreditCardEditor {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        false,
        "addcard",
        false,
        MyValidator {
          "errorsToShow": Array [
            "",
            "",
            "",
            "",
          ],
          "passed": Array [
            false,
            false,
            false,
            true,
          ],
          "requiredError": "Required",
          "validations": Array [
            MyValidation {
              "errorMessage": "Should contain only letters",
              "ruleFormat": /\\(\\[a-zA-Z\\]\\+\\)/i,
            },
            MyValidation {
              "errorMessage": "Should contain 16 digits",
              "ruleFormat": /\\(\\[0-9\\]\\{16\\}\\)/i,
            },
            MyValidation {
              "errorMessage": "Should have date format",
              "ruleFormat": /\\^\\(0\\[1-9\\]\\|1\\[0-2\\]\\)\\\\/\\(\\[0-9\\]\\{2\\}\\)\\$/i,
            },
            MyValidation {
              "errorMessage": "Should contain 3 digits",
              "ruleFormat": /\\(\\[0-9\\]\\{3\\}\\)/i,
            },
          ],
        },
        "",
        "",
        "",
        "123",
        "fas fa-circle ",
        "fas fa-credit-card ",
        "fas fa-calendar-alt ",
        "fas fa-key text-success ",
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {},
      "skip_bound": false,
      "update": [Function],
    },
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
       
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="title mb-3 svelte-180ga2g"
          >
            Add Credit Card
             
          </h5>
           
          <form>
            <div
              class="input-group input-group-lg mb-3"
              data-testid="maindivformHolderName"
            >
              <span
                class="input-group-text"
              >
                <i
                  class="fas fa-circle "
                  data-testid="iconformHolderName"
                />
              </span>
               
              <input
                class="form-control"
                data-testid="inputformHolderName"
                name="formHolderName"
                placeholder="Card Holder Name"
                type="text"
              />
               
            </div>
             
            <div
              class="input-group input-group-lg mb-3"
              data-testid="maindivformCardNumber"
            >
              <span
                class="input-group-text"
              >
                <i
                  class="fas fa-credit-card "
                  data-testid="iconformCardNumber"
                />
              </span>
               
              <input
                class="form-control"
                data-testid="inputformCardNumber"
                name="formCardNumber"
                placeholder="Card's Number"
                type="text"
              />
               
            </div>
             
            <div
              class="row"
            >
              <div
                class="col-lg-7"
              >
                <div
                  class="input-group input-group-lg mb-3"
                  data-testid="maindivformExpirationDate"
                >
                  <span
                    class="input-group-text"
                  >
                    <i
                      class="fas fa-calendar-alt "
                      data-testid="iconformExpirationDate"
                    />
                  </span>
                   
                  <input
                    class="form-control"
                    data-testid="inputformExpirationDate"
                    name="formExpirationDate"
                    placeholder="MM/YY"
                    type="text"
                  />
                   
                </div>
              </div>
               
              <div
                class="col-lg-5"
              >
                <div
                  class="input-group input-group-lg mb-3"
                >
                  <span
                    class="input-group-text"
                  >
                    <i
                      class="fas fa-key text-success  svelte-180ga2g"
                    />
                  </span>
                   
                  <input
                    class="form-control"
                    data-testid="cvv"
                    name="formCvv"
                    placeholder="CVV"
                    type="password"
                  />
                   
                </div>
              </div>
            </div>
             
            <div
              class="d-grid gap-2"
            >
              <button
                class="btn btn-success btn-lg"
                data-testid="submit"
                disabled=""
                type="button"
              >
                <i
                  class="fas fa-database"
                />
                 
                Add
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`check if subscribe triggers the rendering of the new card to edit 1`] = `
Object {
  "component": CreditCardEditor {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        false,
        "editcard",
        true,
        MyValidator {
          "errorsToShow": Array [
            "",
            "",
            "",
            "",
          ],
          "passed": Array [
            false,
            false,
            false,
            false,
          ],
          "requiredError": "Required",
          "validations": Array [
            MyValidation {
              "errorMessage": "Should contain only letters",
              "ruleFormat": /\\(\\[a-zA-Z\\]\\+\\)/i,
            },
            MyValidation {
              "errorMessage": "Should contain 16 digits",
              "ruleFormat": /\\(\\[0-9\\]\\{16\\}\\)/i,
            },
            MyValidation {
              "errorMessage": "Should have date format",
              "ruleFormat": /\\^\\(0\\[1-9\\]\\|1\\[0-2\\]\\)\\\\/\\(\\[0-9\\]\\{2\\}\\)\\$/i,
            },
            MyValidation {
              "errorMessage": "Should contain 3 digits",
              "ruleFormat": /\\(\\[0-9\\]\\{3\\}\\)/i,
            },
          ],
        },
        "holderName",
        "1234561234567890",
        "11/11",
        "123",
        "fas fa-circle ",
        "fas fa-credit-card ",
        "fas fa-calendar-alt ",
        "fas fa-key ",
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        246,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {},
      "skip_bound": false,
      "update": [Function],
    },
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
       
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="title mb-3 svelte-180ga2g"
          >
            Add Credit Card
             
          </h5>
           
          <form>
            <div
              class="input-group input-group-lg mb-3"
              data-testid="maindivformHolderName"
            >
              <span
                class="input-group-text"
              >
                <i
                  class="fas fa-circle "
                  data-testid="iconformHolderName"
                />
              </span>
               
              <input
                class="form-control"
                data-testid="inputformHolderName"
                name="formHolderName"
                placeholder="Card Holder Name"
                type="text"
              />
               
            </div>
             
            <div
              class="input-group input-group-lg mb-3"
              data-testid="maindivformCardNumber"
            >
              <span
                class="input-group-text"
              >
                <i
                  class="fas fa-credit-card "
                  data-testid="iconformCardNumber"
                />
              </span>
               
              <input
                class="form-control"
                data-testid="inputformCardNumber"
                name="formCardNumber"
                placeholder="Card's Number"
                type="text"
              />
               
            </div>
             
            <div
              class="row"
            >
              <div
                class="col-lg-7"
              >
                <div
                  class="input-group input-group-lg mb-3"
                  data-testid="maindivformExpirationDate"
                >
                  <span
                    class="input-group-text"
                  >
                    <i
                      class="fas fa-calendar-alt "
                      data-testid="iconformExpirationDate"
                    />
                  </span>
                   
                  <input
                    class="form-control"
                    data-testid="inputformExpirationDate"
                    name="formExpirationDate"
                    placeholder="MM/YY"
                    type="text"
                  />
                   
                </div>
              </div>
               
              <div
                class="col-lg-5"
              >
                <div
                  class="input-group input-group-lg mb-3"
                >
                  <span
                    class="input-group-text"
                  >
                    <i
                      class="fas fa-key  svelte-180ga2g"
                    />
                  </span>
                   
                  <input
                    class="form-control"
                    data-testid="cvv"
                    name="formCvv"
                    placeholder="CVV"
                    type="password"
                  />
                   
                </div>
              </div>
            </div>
             
            <div
              class="d-grid gap-2"
            >
              <button
                class="btn btn-success btn-lg"
                data-testid="submit"
                disabled=""
                type="button"
              >
                <i
                  class="fas fa-database"
                />
                 
                Add
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
